concordium-smart-contract-testing/src/impls.rs

```diff
@@ -1056,11 +1223,277 @@ impl Chain {
+            Some(data) => Ok(data),
+        }
+    }
+
     /// Return the current microCCD per euro exchange rate.
     pub fn micro_ccd_per_euro(&self) -> ExchangeRate { self.parameters.micro_ccd_per_euro }
 
     /// Return the current euro per energy exchange rate.
     pub fn euro_per_energy(&self) -> ExchangeRate { self.parameters.euro_per_energy }
+
+    /// Return the current block time.
+    pub fn block_time(&self) -> Timestamp { self.parameters.block_time }
+
+    /// Return the "sticky block", i.e. the block that is saved and will be used
+    /// by default for future queries.
+    ///
+    /// See [`set_block_time_via_external_node`][Chain::set_block_time_via_external_node] for a more detailed explanation.
+    pub fn sticky_block(&self) -> Option<BlockHash> {
+        self.external_node_connection
+            .as_ref()
+            .and_then(|conn| conn.sticky_block)
+    }
+}
+
+impl ExternalNodeConnection {
+    /// Execute an async task with the [`sdk::v2::Client`].
+    ///
+    /// The task will be performed using one of the following three blocks:
+    ///  1. If a `block` hash is provided, it will be used.
+    ///  2. Otherwise, if `self.sticky_block` is `Some`, it will be used.
+    ///  3. Lastly, if neither are present, the last final block is used.
+    ///
+    /// If `self.sticky_block` is `None`, then it will be set to the block used
+    /// in cases 1 and 3.
+    ///
+    /// If the task takes longer than [`EXTERNAL_NODE_TIMEOUT_DURATION`] then
+    /// the connection times out and an [`Err(ExternalNodeError::Timeout)`] is
+    /// returned.
+    fn with_client<T, F, Fut>(
+        &mut self,
+        block: Option<BlockHash>,
+        f: F,
+    ) -> Result<T, ExternalNodeError>
+    where
+        F: FnOnce(sdk::v2::BlockIdentifier, sdk::v2::Client) -> Fut,
+        Fut: Future<Output = Result<(BlockHash, T), ExternalNodeError>>, {
+        // Get the block identifier in one of the three ways.
+        let block_identifier = match (block, self.sticky_block) {
+            // Case 1
+            (Some(block), _) => sdk::v2::BlockIdentifier::Given(block),
+            // Case 2
+            (_, Some(sticky_block)) => sdk::v2::BlockIdentifier::Given(sticky_block),
+            // Case 3
+            _ => sdk::v2::BlockIdentifier::LastFinal,
+        };
+        // Clone the client so it can be moved to the async block.
+        let client = self.client.clone();
+        // Run the future and timeout if it takes too long.
+        let res = self.runtime.block_on(async move {
+            timeout(EXTERNAL_NODE_TIMEOUT_DURATION, f(block_identifier, client))
+                .await
+                .map_err(|_| ExternalNodeError::Timeout)?
+        });
+        // Extract the return value and maybe set the stick block.
+        match res {
+            Ok((block_hash, return_value)) => {
+                if self.sticky_block.is_none() {
+                    self.sticky_block = Some(block_hash)
+                }
```

Comment:
I would probably expect more the behaviour that "sticky" block is the last block that was used.

So it is overridden even if it is set already, and I supply a custom block.